seq_tools/validation/c660_metadata_in_bam_rg_header.py

```diff
@@ -0,0 +1,163 @@
+
+            header_array = header.rstrip().split('\n')
+
+            rgs_in_bam = {}
+            for line in header_array:
+                if not line.startswith("@RG"):
+                    continue
+
+                rg = {}
+                for kv in line.split('\t'):
+                    if ':' not in kv:  # like '@RG'
+                        continue
+                    k, v = kv.split(':')[0], ':'.join(kv.split(':')[1:])
+                    rg[k] = v
+
+                # all RG IDs from one BAM must be unique, this has been checked in c610_rg_id_in_bam
+                rgs_in_bam[rg['ID']] = rg
+
+            all_rgs[f] = rgs_in_bam
+
+        # now comparing read group info in metadata with all_rgs from BAM header collected above
+        mismatches = defaultdict(dict)
+        for rg in self.metadata['read_groups']:
+            seq_file = rg['file_r1']
+            if not seq_file.endswith('.bam'):  # skip if not BAM
+                continue
+
+            read_group_id = rg['submitter_read_group_id']
+            if rg.get('read_group_id_in_bam'):
+                read_group_id = rg['read_group_id_in_bam']
+
+            for m in field_mapping:
+                field_in_bam, ex_or_rg, field_in_meta = m
+
+                value_in_bam = all_rgs.get(seq_file, {}).get(read_group_id, {}).get(field_in_bam)
+                if ex_or_rg == 'rg':
+                    value_in_meta = rg.get(field_in_meta)
+                elif ex_or_rg == 'experiment':
+                    value_in_meta = self.metadata['experiment'][field_in_meta]
+
+                if value_in_bam and value_in_bam != value_in_meta:
+                    if mismatches.get(seq_file, {}).get(read_group_id) is None:
+                        mismatches[seq_file][read_group_id] = []
+
+                    mismatches[seq_file][read_group_id].append(
+                        "%s: %s vs %s" % (field_in_bam, value_in_bam, value_in_meta)
+                    )
+                elif value_in_bam:  # matches
+                    self.logger.info("[%s] BAM: %s, RG: %s, field: %s, value: %s matches what is in metadata." %
+                                     (self.checker, seq_file, read_group_id, field_in_bam, value_in_bam))
+
+        if mismatches:
+            mismatches_strings = []
+            for f in mismatches:  # file
+                for rg in mismatches[f]:  # rg
+                    mismatches_strings.append("[BAM %s @RG %s: (%s)]" % (f, rg, ", ".join(mismatches[f][rg])))
+
+            self.status = 'WARNING'
+            message = "Information (excluding ID and SM which are validated elsewhere) in BAM @RG header does NOT match " \
+                "experiment/read group info in metadata JSON. NOTE that information is the metadata JSON document will be " \
+                "kept and used in ICGC ARGO uniform analysis while unmatched info in BAM header will be lost. Details of " \
+                "the difference: %s" % "; ".join(mismatches_strings)
+
```

Comment:
Had a very trivial thought here: can we improve the 'readability' a bit by storing  `message` as an array. Like:
```json
{
  "message":[
    "Information (excluding ID and SM which are validated elsewhere) in BAM @rg header does NOT match experiment/read group info in metadata JSON.",
    "NOTE: that information is the metadata JSON document will be kept and used in ICGC ARGO uniform analysis while unmatched info in BAM header will be lost. ",
    "Details of the difference: [BAM test_rg_3.bam @rg C0HVY.2: (CN: BI vs EXT)]; [BAM test_rg_3.bam @rg D0RE2_1': (CN: BI vs EXT)]; [BAM test_rg_3.bam @rg D0RH0:2'': (CN: BI vs EXT)]"
  ]
}
```

Reply:
one constraint we have is that this message will be output to a log file which is plain text. Maybe we can keep the current message field, but introduce the other as list for details, this details can be output to log file as individual lines.